---
title: "Msmc Riz"
author: "Philippe Cubry"
date: "9 août 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r functions and libraries}
estimate_mode <- function(x) {
  d <- density(x)
  d$x[which.max(d$y)]
}

library(cowplot)

```

# Introduction

This report highlight the methods and results of an msmc analysis built upon the African rice genomic data. 

## Generating the needed files

### Masks
In order to use msmc, one should generate the input file using a VCF and a "mask" file that list the positions that should be mapped on the ref sequence. Using msmc tools, it might be pretty easy to generate those file, starting from a bam for each unique individual considered in the study.
To test the approach, we will use genotypes from a given defined group (see the "glaberrima group definition" report for more information). We will begin by using genotypes from group 5, i.e. EA/FP/FT/GC/GP/HE/MI/NG. Corresponding codes in the .bam  


Label | BAM code
---- | ------
EA | 1.2.M.D
FP | 1.F.1T
FT | 1.J.29
GC | 1.2.K
GP | 1.D.V
HE | 1.4.G
MI | 1.4
NG | 1.1.2  

Log in node2  

* ssh node2  

Make temp directory  

* mkdir /scratch/cubry/ 

Copy necessary .bam file (example with chr 3)  

* scp cubry@nas:/data2/projects/irigin/BAM-ALL-SAMPLES/BAM-GLAB/OG-all-MSU7-CHR3.bam cubry@node2:/scratch/cubry/  

Log in with SGE  

* qrsh -q *@node2  

Extract individual bam files (example for chr. 1)  

* for i in {1.A.U.1B,1.A.19,1.C.1H,1.9.R,1.F.X,1.5.K,1.B.18,1.5.L,1.J.25,1.B}; do echo $i > $i.txt; samtools view -R $i.txt -b OG-all-MSU7-CHR1.bam > OG-all-MSU7-CHR1.bam.only$i.bam; done  

Use the bamCaller.py function with a qsub command, arguments as follow  

* For samtools mpileup
** -q 20 -Q 20 -C 50 -u -f /data2/projects/africrop_models/msmc/ref_riz/MSU7.fa $1 (bam file from the qsub command)  

* For bcftools arguments from the msmc tools manual  

* For bamCaller.py default arguments, except mean coverage fixed to 37 (Cécile pers comm) and output file to $1.mask.bed.gz

We should use additional groups from the border of the sampling, e.g. group 7 and 3.
Genotypes from group 3:

Label | BAM code
----- | -------
EI | 1.A.U.1B
FI | 1.A.19
FL | 1.C.1H
GK | 1.9.R
GR | 1.F.X
HF | 1.5.K
HM | 1.B.18
KF | 1.5.L
LA | 1.J.25
MR | 1.B

Genotypes from group 7:

Label | BAM code
----- | -------
EN | 1.E.Y.1R
FC | 1.4.L
FS | 1.I.25
GB | 1.1.J
HB | 1.1.4.5
HC | 1.2.8
IM | 1.B.U
IQ | 1.E.X
KM | 1.B.19
LB | 1.1.5
LR | 1.B.V
MG | 1.2.6
MN | 1.8
MP | 1.9

Now, we just want to rename the files according to their individual name, e.g. 1.8 will become MN

\#!/bin/bash

for file in *1.8*; do

mv $file ${file/1.8/MN}

done



### VCF files
To better manage files, we will extract only useful information from the initial VCF provided by Christine (the ones after applying all filters). We'll use vcftools to extract only the individuals we're interested in and to suppress all non-needed information fields and keeping only the glaberrima genotypes

* vcftools --vcf <VCFfile> --keep keep_all_glab.txt --recode --out <output prefix>


We then transform the resulting VCF in a VCF-like file containing haploidized individuals with a home-made R script, using the following rules:  

* if genotype was 0/0, replaced by 0
* if genotype was 1/1, replaced by 1
* if genotype was 0/1, randomly replaced by 0 OR 1
* if genotype was ./., replaced by .

From these haploid VCF-like files, we will now generate fake diploids in VCF format. For this purpose, we first define the couples of individuals to be merged, then merge the genotypes with the following rules:  

* if both genotypes were 0, then write 0|0
* if both genotypes were 1, then write 1|1
* if the first genotype is 0 and the other 1, then write 0|1
* if the first genotype is 1 and the other 0, then write 1|0
* if one genotype is missing, write .|.

We then write in the output file only the polymorphic (i.e. 0|1, 1|0) and monomorphic sites for alternative allele (1|1). Missing (.|.) positions were saved onto another vcf file.



home-made scripts were called "make_haplo_vcf.R" and "make_fake_diplo_vcf.R"  


### List positions discarded after filters
Various filters have been applied in order to generate the final vcf files. We'll have to filter out the discarded positions from the analysis. For that we will have to generate bed files listing these positions.  

Example for chr1

* bedtools intersect -a OgOb-all-MSU7-CHR1.vcf -b OgOb-all-MSU7-CHR1.SNPEFF.FILTERED.NA.225Homoz.minmaxAllele2.recode-AF1bis.vcf -v > CHR1_filtered_pos.vcf  

We then want to convert the filtered positions into a bed file using ad hoc script:  

sed -e 's/chr//' file.vcf | awk '{OFS="\\t"; if (!/^#/){print $1,$2-1,$2}}'

```{r correct code without character escapment, eval=FALSE, include=FALSE}
sed -e 's/chr//' file.vcf | awk '{OFS="\t"; if (!/^#/){print $1,$2-1,$2}}'
```

We applied the same process to the vcf containing the missing files generated in the previous chapter.

### Compute input files for msmc
For this purpose, we'll use the msmc-tools script 
generate_multihetsep.py
We'll use only one mask (the one of genotype GB for glaberrima and the one of DK for barthii) and one negative mask per chromosome


# For barthii
We first haploidized barthii genotypes

* vcftools --vcf <VCFfile> --keep keep-barth-georef.txt --recode --out <output prefix>

We made the same keeping all genotypes known as not being adventices

* vcftools --vcf <VCFfile> --keep keep-barth-ALL-withtAdventices.txt --recode --out <output prefix>


We then sampled 16 haplotypes to make 8 false diploids (using same scripts and procedure as for glaberrima)

The 16 sampled genotypes were:

Label | BAM code
----- | -------
BG | 1.7
BN | 1.D
BP | 1.E
BQ | 1.F
BR | 1.G
BS | 1.H
CE | 1.4.L
CN | 1.C.T
CT | 1.H.Y
CV | 1.I.1I
DC | 1.3.9
DF | 1.6.I
DI | 1.9.R
DK | 1.A.U
DN | 1.D.13
NK | NK


## Extraction from bam files
We then extracted the sampled individuals from the bam files in order to compute masks.

We then used a mean sequencing depth of 33x for mpileup (Cecile data)

#Running msmc
We then run msmc once for each species:

* On 16 haplotypes (false diploids)

* On the 8 pseudo-diploids assumed (PSMC')

##Bootstraping
A built-in script from msmc-tools allow the creation of bootstraped input files for msmc. We generated (only for the MSMC analysis, not the PSMC') 20 bootstraped input files as follow (example for glaberrima)

./multihetsep_bootstrap.py -n 20 -s 2000000 --chunks_per_chromosome 20 --nr_chromosomes 12 16haplos_bootstraped /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr1_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr2_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr3_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr4_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr5_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr6_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr7_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr8_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr9_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr10_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr11_16haplos_msmc_input /data2/projects/africrop_models/msmc/msmc_inputFiles/16haplos/chr12_16haplos_msmc_input

# Convergence between msmc and stairway
We detected a difference in the mutation rate assumed for each method. This should result from the initial approximation made for the stairway analysis when trying to estimate the length of the sequence that gave the observed SFS (based on SNP density from the total genome length). Here we calculate the size of the mappable genome based on the masks generated for the msmc analysis. We chose the genotypes "MG", "GR", "LR" and "GB" to compute this size.
Script estimate_mappable_size.R
Size of MG mappable genome = 197407227
Size of GR mappable genome = 213097435
Size of LR mappable genome = 203362793
Size of GB mappable genome = 207482589
Mean value : 205337511

For Barthii, we used genotypes "BQ","CE","CV","DK" and "DN".
Size of BQ mappable genome = 179137944
Size of CE mappable genome = 177543127
Size of CV mappable genome = 213235236
Size of DK mappable genome = 200482235
Size of DN mappable genome = 218350399
Mean value : 197749788

# Resulting plot for O glaberrima
Below is presented the results of the three analysis on the same plot, assuming a mutation rate of 6.5e-9. For Stairway, 200 bootstraps were made as recommended by the authors. For PSMC', we used repetitions as a proxi for the variance in the estimation. For MSMC; here are presented the result of the main analyisis (thick blue line) as well as the results for 6 bootstraped datasets (dashed blue lines).
For reminder, a vertical line was drawn at 2700 generations ago, corresponding roughly to the oldest known fosill evidence of domesticated form of african rice.

```{r, echo=FALSE,message=FALSE}
msmc_16 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos.final.txt",header=TRUE)
msmc_16.b1 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped_1.final.txt",header=TRUE)
msmc_16.b12 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped_12.final.txt",header=TRUE)
msmc_16.b4 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped_4.final.txt",header=TRUE)
msmc_16.b7 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped_7.final.txt",header=TRUE)
msmc_16.b9 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped_9.final.txt",header=TRUE)
msmc_16.b10 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_16haplos_bootstraped10.final.txt",header=TRUE)
msmc.2haplo.1 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_FI_KT.final.txt",header=TRUE)
msmc.2haplo.2 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_GB_MC.final.txt",header=TRUE)
msmc.2haplo.3 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_GV_KH.final.txt",header=TRUE)
msmc.2haplo.4 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_KE_MD.final.txt",header=TRUE)
msmc.2haplo.5 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_LM_FQ.final.txt",header=TRUE)
msmc.2haplo.6 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_MB_HN.final.txt",header=TRUE)
msmc.2haplo.7 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_MS_EC.final.txt",header=TRUE)
msmc.2haplo.8 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_2haplos_NB_GM.final.txt",header=TRUE)


og <- read.table("/Users/cubry/Documents/Riz_analysis/stairway/Og_msmcbased_allsnps/theta/og_unfolded_sfs_allSNPs.65_bootAll.out",skip=5,header=TRUE)
testFinal <- read.table("/Users/cubry/Documents/Riz_analysis/stairway/Og_msmcbased_allsnps_stairV2/cult_rice_msmc_scaled_all_snps.final.summary",header=TRUE)

library(ggplot2)
library(cowplot)

mu <- 6.5e-9
gen <- 1
msmc_16.scaled <- NULL
msmc_16.scaled$year<- msmc_16$left_time_boundary/(mu*gen)
msmc_16.scaled$Ne<-  (1/msmc_16$lambda)/(2*mu)
msmc_16.scaled <- as.data.frame(msmc_16.scaled)

msmc_16.b1.scaled  <- NULL
msmc_16.b1.scaled$year<- msmc_16.b1$left_time_boundary/(mu*gen)
msmc_16.b1.scaled$Ne<-  (1/msmc_16.b1$lambda)/(2*mu)
msmc_16.b1.scaled <- as.data.frame(msmc_16.b1.scaled)

msmc_16.b12.scaled  <- NULL
msmc_16.b12.scaled$year<- msmc_16.b12$left_time_boundary/(mu*gen)
msmc_16.b12.scaled$Ne<-  (1/msmc_16.b12$lambda)/(2*mu)
msmc_16.b12.scaled <- as.data.frame(msmc_16.b12.scaled)

msmc_16.b4.scaled  <- NULL
msmc_16.b4.scaled$year<- msmc_16.b4$left_time_boundary/(mu*gen)
msmc_16.b4.scaled$Ne<-  (1/msmc_16.b4$lambda)/(2*mu)
msmc_16.b4.scaled <- as.data.frame(msmc_16.b4.scaled)

msmc_16.b7.scaled  <- NULL
msmc_16.b7.scaled$year<- msmc_16.b7$left_time_boundary/(mu*gen)
msmc_16.b7.scaled$Ne<-  (1/msmc_16.b7$lambda)/(2*mu)
msmc_16.b7.scaled <- as.data.frame(msmc_16.b7.scaled)

msmc_16.b9.scaled  <- NULL
msmc_16.b9.scaled$year<- msmc_16.b9$left_time_boundary/(mu*gen)
msmc_16.b9.scaled$Ne<-  (1/msmc_16.b9$lambda)/(2*mu)
msmc_16.b9.scaled <- as.data.frame(msmc_16.b9.scaled)

msmc_16.b10.scaled  <- NULL
msmc_16.b10.scaled$year<- msmc_16.b10$left_time_boundary/(mu*gen)
msmc_16.b10.scaled$Ne<-  (1/msmc_16.b10$lambda)/(2*mu)
msmc_16.b10.scaled <- as.data.frame(msmc_16.b10.scaled)

msmc.2haplo.1.scaled <- NULL
msmc.2haplo.1.scaled$year<- msmc.2haplo.1$left_time_boundary/(mu*gen)
msmc.2haplo.1.scaled$Ne<-  (1/msmc.2haplo.1$lambda)/(2*mu)
msmc.2haplo.1.scaled <- as.data.frame(msmc.2haplo.1.scaled)

msmc.2haplo.2.scaled  <- NULL
msmc.2haplo.2.scaled$year<- msmc.2haplo.2$left_time_boundary/(mu*gen)
msmc.2haplo.2.scaled$Ne<-  (1/msmc.2haplo.2$lambda)/(2*mu)
msmc.2haplo.2.scaled <- as.data.frame(msmc.2haplo.2.scaled)

msmc.2haplo.3.scaled  <- NULL
msmc.2haplo.3.scaled$year<- msmc.2haplo.3$left_time_boundary/(mu*gen)
msmc.2haplo.3.scaled$Ne<-  (1/msmc.2haplo.3$lambda)/(2*mu)
msmc.2haplo.3.scaled <- as.data.frame(msmc.2haplo.3.scaled)

msmc.2haplo.4.scaled  <- NULL
msmc.2haplo.4.scaled$year<- msmc.2haplo.4$left_time_boundary/(mu*gen)
msmc.2haplo.4.scaled$Ne<-  (1/msmc.2haplo.4$lambda)/(2*mu)
msmc.2haplo.4.scaled <- as.data.frame(msmc.2haplo.4.scaled)

msmc.2haplo.5.scaled <- NULL
msmc.2haplo.5.scaled$year<- msmc.2haplo.5$left_time_boundary/(mu*gen)
msmc.2haplo.5.scaled$Ne<-  (1/msmc.2haplo.5$lambda)/(2*mu)
msmc.2haplo.5.scaled <- as.data.frame(msmc.2haplo.5.scaled)

msmc.2haplo.6.scaled <- NULL
msmc.2haplo.6.scaled$year<- msmc.2haplo.6$left_time_boundary/(mu*gen)
msmc.2haplo.6.scaled$Ne<-  (1/msmc.2haplo.6$lambda)/(2*mu)
msmc.2haplo.6.scaled <- as.data.frame(msmc.2haplo.6.scaled)

msmc.2haplo.7.scaled <- NULL
msmc.2haplo.7.scaled$year<- msmc.2haplo.7$left_time_boundary/(mu*gen)
msmc.2haplo.7.scaled$Ne<-  (1/msmc.2haplo.7$lambda)/(2*mu)
msmc.2haplo.7.scaled <- as.data.frame(msmc.2haplo.7.scaled)

msmc.2haplo.8.scaled <- NULL
msmc.2haplo.8.scaled$year<- msmc.2haplo.8$left_time_boundary/(mu*gen)
msmc.2haplo.8.scaled$Ne<-  (1/msmc.2haplo.8$lambda)/(2*mu)
msmc.2haplo.8.scaled <- as.data.frame(msmc.2haplo.8.scaled)

Ne_infered.plot <- ggplot() +
  geom_step(data = msmc_16.b1,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b12,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b4,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b7,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b9,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b10,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "MSMC"),size=1.5)+
  geom_step(data = msmc.2haplo.1,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.2,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.3,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.4,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.5,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.6,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.7,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.8,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = og,aes(x = year, y = Ne_median,colour="Stairway"),alpha=0.5,size=1.5)+
  geom_ribbon(data = og,aes(x = year, ymin = Ne_2.5., ymax = Ne_97.5.),alpha=0.2)+
  geom_step(data = testFinal,aes(x = year, y = Ne_median,colour="Stairway v2"),alpha=0.5,size=1.5)+
  geom_ribbon(data = testFinal,aes(x = year, ymin = Ne_2.5., ymax = Ne_97.5.),alpha=0.2,fill="orange")+
  scale_y_log10() +scale_x_log10()+ 
  scale_colour_manual(name="", values = c("blue","green", "black","orange")) +
  xlab("Log10 of generations")+ylab("Log 10 of effective size")


Ne_infered.combined_plot <- ggplot() +
  geom_step(data = msmc_16.b1.scaled[msmc_16.b1.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b12.scaled[msmc_16.b12.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b4.scaled[msmc_16.b4.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b7.scaled[msmc_16.b7.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b9.scaled[msmc_16.b9.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b10.scaled[msmc_16.b10.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.scaled[msmc_16.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),size=1.5)+
  geom_step(data =  msmc.2haplo.1.scaled[msmc.2haplo.1.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.2.scaled[msmc.2haplo.2.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.3.scaled[msmc.2haplo.3.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data =msmc.2haplo.4.scaled[msmc.2haplo.4.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.5.scaled[msmc.2haplo.5.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data =msmc.2haplo.6.scaled[msmc.2haplo.6.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.7.scaled[msmc.2haplo.7.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.8.scaled[msmc.2haplo.8.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = og[og$year<1500,],aes(x = year, y = Ne_median,colour="Stairway"),alpha=0.5,size=1.5)+
  geom_ribbon(data = og[og$year<1500,],aes(x = year, ymin = Ne_2.5., ymax = Ne_97.5.),alpha=0.2)+
  scale_y_log10() +scale_x_log10()+ 
  scale_colour_manual(name="", values = c("blue","green", "black","orange")) +
  xlab("Log10 of generations")+ylab("Log 10 of effective size")

# save_plot(Ne_infered.plot,filename = "Ne_infered.plot.pdf",base_height = 10,base_aspect_ratio = 1.2)

# save_plot(Ne_infered.combined_plot,filename = "Ne_infered.combined_plot.pdf",base_height = 10,base_aspect_ratio = 1.2)
Ne_infered.plot
Ne_infered.combined_plot


Ne_infered.decrease_plot <- ggplot() +
  geom_step(data = msmc_16.b1.scaled[msmc_16.b1.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b12.scaled[msmc_16.b12.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b4.scaled[msmc_16.b4.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b7.scaled[msmc_16.b7.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b9.scaled[msmc_16.b9.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.b10.scaled[msmc_16.b10.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),linetype="dashed")+
  geom_step(data = msmc_16.scaled[msmc_16.scaled$year>1000,],aes(x = year, y = Ne,colour= "MSMC"),size=1.5)+
  geom_step(data =  msmc.2haplo.1.scaled[msmc.2haplo.1.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.2.scaled[msmc.2haplo.2.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.3.scaled[msmc.2haplo.3.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data =msmc.2haplo.4.scaled[msmc.2haplo.4.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.5.scaled[msmc.2haplo.5.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data =msmc.2haplo.6.scaled[msmc.2haplo.6.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.7.scaled[msmc.2haplo.7.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  geom_step(data = msmc.2haplo.8.scaled[msmc.2haplo.8.scaled$year>1000,],aes(x = year, y = Ne,colour= "PSMC'"),alpha=0.5,size=1.5)+
  scale_y_log10() + xlim(2000,20000)+ 
  scale_colour_manual(name="", values = c("blue","green", "black","orange")) +
  xlab("generations")+ylab("Log 10 of effective size")


# save_plot(Ne_infered.plot,filename = "Ne_infered.plot.pdf",base_height = 10,base_aspect_ratio = 1.2)

# save_plot(Ne_infered.combined_plot,filename = "Ne_infered.combined_plot.pdf",base_height = 10,base_aspect_ratio = 1.2)

# save_plot(Ne_infered.decrease_plot,filename = "Ne_infered.decrease_plot.pdf",base_height = 10,base_aspect_ratio = 1.2)

Ne_infered.plot
Ne_infered.combined_plot
Ne_infered.decrease_plot
```


We below plot the results of PSMC' analysis on barthii genotypes. For comparison, we redraw in this graph the results of PSMC' on glaberrima. We also included the results of the stairway analysis based on the full set of available SNP for the georeferenced barthii genotypes.

```{r msmc barthii random,echo=FALSE,message=FALSE}
msmc.barthii.2haplo.1 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_BG_CV.final.txt",header=TRUE)
msmc.barthii.2haplo.2 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_BN_BR.final.txt",header=TRUE)
msmc.barthii.2haplo.3 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_BP_DF.final.txt",header=TRUE)
msmc.barthii.2haplo.4 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_BQ_DK.final.txt",header=TRUE)
msmc.barthii.2haplo.5 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_BS_CE.final.txt",header=TRUE)
msmc.barthii.2haplo.6 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_CN_DI.final.txt",header=TRUE)
msmc.barthii.2haplo.7 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_CT_DC.final.txt",header=TRUE)
msmc.barthii.2haplo.8 <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/msmc_barthii_2haplos_NK_DN.final.txt",header=TRUE)



library(ggplot2)
library(cowplot)

mu <- 6.5e-9
gen <- 1


ggplot() +
  geom_step(data = msmc.barthii.2haplo.1,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.2,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.3,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.4,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.5,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.6,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.7,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.barthii.2haplo.8,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barth"))+
  geom_step(data = msmc.2haplo.1,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.2,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.3,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.4,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.5,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.6,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.7,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  geom_step(data = msmc.2haplo.8,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),size=1)+
  scale_y_log10() + scale_x_log10() +
  scale_colour_manual(name="", values = c("black","red","gray","green","pink","brown","orange","cyan","blue","purple")) +
  xlab("Log10 of generations")+ylab("Log 10 of effective size") +
  ggtitle("Inference of past demography\n Mutation rate of 6.5e-9")

```

#29 september 2016 - Barthii update
To infer Ne for Barthii, we chose to adopt a structure-based approach. We used the outcomes of the snmf analysis at K=6 from Anne Celine to define genetic clusters. We used a 0.5 threshold for the consideration of assignment to a cluster. This leaved us with 5 groups with more than two genotypes.
We created lists of genotypes for each group, sorted by decreasing assignment to the considered group.
We then make fake diploids for genotypes taken in these groups.


```{r msmc barthii per group,echo=FALSE,message=FALSE}
msmc.barthii.2haplo.AC_AN <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AC_AN_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AE_AF <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AE_AF_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AH_AD <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AH_AD_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AI_AM <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AI_AM_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AL_AK <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AL_AK_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AP_AG <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AP_AG_G4.final.txt",header=TRUE)
msmc.barthii.2haplo.AR_NR <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AR_NR_G3.final.txt",header=TRUE)
msmc.barthii.2haplo.AS_NN <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AS_NN_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.AV_CH <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_AV_CH_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.CE_DK <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CE_DK_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.CF_DA <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CF_DA_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.CG_BQ <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CG_BQ_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.BC_CC <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BC_CC_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BE_NP <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BE_NP_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BF_NQ <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BF_NQ_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BH_BG <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BH_BG_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BI_CN <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BI_CN_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BK_CS <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BK_CS_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BM_CB <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BM_CB_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BP_BL <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BP_BL_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BV_BD <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BV_BD_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.CI_CR <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CI_CR_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.CK_CP <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CK_CP_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.CQ_CL <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CQ_CL_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.BR_CD <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_BR_CD_G5.final.txt",header=TRUE)

msmc.barthii.2haplo.CV_DI <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_CV_DI_G5.final.txt",header=TRUE)

msmc.barthii.2haplo.DC_DB <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DC_DB_G5.final.txt",header=TRUE)

msmc.barthii.2haplo.DD_DN <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DD_DN_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.DF_AQ <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DF_AQ_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.DM_DG <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DM_DG_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.NS_S <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_NS_S_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.R_CT <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_R_CT_G3.final.txt",header=TRUE)

msmc.barthii.2haplo.DE_BA <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DE_BA_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.DH_BS <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DH_BS_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.DL_CM <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_DL_CM_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.NT_BB <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_NT_BB_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.NV_CA <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_NV_CA_G2.final.txt",header=TRUE)

msmc.barthii.2haplo.NK_AB <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_NK_AB_G5.final.txt",header=TRUE)

msmc.barthii.2haplo.T_BN <- read.table("/Users/cubry/Documents/Riz_analysis/msmc/results/psmc_T_BN_G1.final.txt",header=TRUE)

library(ggplot2)
library(cowplot)

mu <- 6.5e-9
gen <- 1


psmc_wild <- ggplot() +
  geom_step(data = msmc.barthii.2haplo.AC_AN,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  #geom_step(data = msmc.barthii.2haplo.AE_AF,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AH_AD,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AI_AM,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AL_AK,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AP_AG,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AR_NR,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AS_NN,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.AV_CH,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CE_DK,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CF_DA,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CG_BQ,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BC_CC,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BE_NP,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BF_NQ,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BH_BG,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  #geom_step(data = msmc.barthii.2haplo.BI_CN,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BK_CS,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BM_CB,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BP_BL,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.BV_BD,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CI_CR,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CK_CP,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CQ_CL,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  #geom_step(data = msmc.barthii.2haplo.BR_CD,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.CV_DI,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DC_DB,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DD_DN,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  #geom_step(data = msmc.barthii.2haplo.DF_AQ,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DM_DG,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.NS_S,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.R_CT,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DE_BA,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DH_BS,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.DL_CM,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.NT_BB,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.NV_CA,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.NK_AB,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.barthii.2haplo.T_BN,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' barthii"),alpha=0.5,size=0.5)+
   geom_step(data = msmc.2haplo.1,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.2,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.3,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.4,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.5,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.6,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.7,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_step(data = msmc.2haplo.8,aes(x = left_time_boundary/(mu*gen), y = (1/lambda)/(2*mu),colour= "PSMC' glab"),alpha=0.5,size=0.5)+
  geom_ribbon(aes(x=c(2400,2800),ymin = 0,ymax=1e7),alpha=0.2)+
  geom_text(aes(x=2800,y=6e2,hjust=-0.02,vjust=0,label="400-800 BC - Most ancient evidence \n of african rice domesticated form"))+
  scale_y_log10() + scale_x_log10() +
  scale_colour_manual(name="", values = c("purple","green")) +
  xlab("Log10 of generations")+ylab("Log 10 of effective size")

```

```{r calculate density of latest growth}
t.wild<- NULL
for(g in c("msmc.barthii.2haplo.AC_AN","msmc.barthii.2haplo.AH_AD","msmc.barthii.2haplo.AI_AM",
           "msmc.barthii.2haplo.AL_AK","msmc.barthii.2haplo.AP_AG","msmc.barthii.2haplo.AR_NR",
           "msmc.barthii.2haplo.AS_NN","msmc.barthii.2haplo.AV_CH","msmc.barthii.2haplo.CE_DK",
           "msmc.barthii.2haplo.CF_DA","msmc.barthii.2haplo.CG_BQ","msmc.barthii.2haplo.BC_CC",
           "msmc.barthii.2haplo.BE_NP","msmc.barthii.2haplo.BF_NQ","msmc.barthii.2haplo.BH_BG",
           "msmc.barthii.2haplo.BK_CS","msmc.barthii.2haplo.BM_CB","msmc.barthii.2haplo.BP_BL",
           "msmc.barthii.2haplo.BV_BD","msmc.barthii.2haplo.CI_CR","msmc.barthii.2haplo.CK_CP",
           "msmc.barthii.2haplo.CQ_CL",
           #"msmc.barthii.2haplo.BR_CD","msmc.barthii.2haplo.BI_CN",
           #"msmc.barthii.2haplo.AE_AF","msmc.barthii.2haplo.DF_AQ",
           "msmc.barthii.2haplo.CV_DI","msmc.barthii.2haplo.DC_DB","msmc.barthii.2haplo.DD_DN",
           "msmc.barthii.2haplo.DM_DG","msmc.barthii.2haplo.NS_S","msmc.barthii.2haplo.R_CT",
           "msmc.barthii.2haplo.DE_BA","msmc.barthii.2haplo.DH_BS","msmc.barthii.2haplo.DL_CM",
           "msmc.barthii.2haplo.NT_BB","msmc.barthii.2haplo.NV_CA","msmc.barthii.2haplo.NK_AB",
           "msmc.barthii.2haplo.T_BN")){
  g <- get(g)
  g$left_time_boundary<-g$left_time_boundary/(mu*gen)
  g$lambda <- (1/g$lambda)/(2*mu)
  t.wild<-c(t.wild,(g$left_time_boundary[3]))
}

t.cult<- NULL
for(g in c("msmc.2haplo.1","msmc.2haplo.2",
           "msmc.2haplo.3","msmc.2haplo.4",
           "msmc.2haplo.5","msmc.2haplo.6",
           "msmc.2haplo.7","msmc.2haplo.8")){
  g <- get(g)
  g$left_time_boundary<-g$left_time_boundary/(mu*gen)
  g$lambda <- (1/g$lambda)/(2*mu)
  t.cult<-c(t.cult,(g$left_time_boundary[3]))
}

par(mfrow=c(2,1))
plot(density(t.wild),xlim=c(0,8000))
abline(v=estimate_mode(t.wild))

plot(density(t.cult),xlim=c(0,8000))
abline(v=estimate_mode(t.cult))

density_wild <- ggplot()  +
            #Plot the density graphs
            geom_density(aes(t.wild,fill="Barthii post-bottleneck\ngrowth density"))+
            #labels
            labs(x="Log10 of generations") +
            #Define the theme  
            #theme_bw() +
            scale_fill_manual(values = "purple")+
            scale_x_log10()+  
            theme(panel.grid=element_blank(),legend.title=element_blank())
density_cult <- ggplot()  +
            #Plot the density graphs
            geom_density(aes(t.cult,fill="Glab post-bottleneck\ngrowth density"))+
            #labels
            labs(x="") +
            #Define the theme  
            #theme_bw() +
            scale_fill_manual(values = "green")+
            scale_x_log10()+
            theme(panel.grid=element_blank(),legend.title=element_blank())

x.min <- 10^(ggplot_build(psmc_wild))$layout$panel_ranges[[1]]$x.range[1]
x.max <- 10^(ggplot_build(psmc_wild))$layout$panel_ranges[[1]]$x.range[2]

psmc_wild <- psmc_wild + scale_x_log10(limits=c(20,x.max+1))
density_cult <- density_cult + scale_x_log10(limits=c(20,x.max+1))
density_wild <- density_wild + scale_x_log10(limits=c(20,x.max+1))

save_plot(plot_grid(psmc_wild + labs(x="") ,
                    density_cult + geom_ribbon(aes(x=c(2400,2800),ymin = 0,ymax=5),alpha=0.2)+
  geom_text(aes(x=2800,y=1,hjust=-0.02,vjust=0,label="400-800 BC - Most ancient evidence \n of african rice domesticated form"))+ geom_vline(aes(xintercept=estimate_mode(t.cult)))+ geom_text(aes(x=estimate_mode(t.cult)+10,y = 4.5,label="Mode of the distribution = 1847",hjust=1.05,vjust=0)),
                    density_wild +  geom_ribbon(aes(x=c(2400,2800),ymin = 0,ymax=3.5),alpha=0.2)+
  geom_text(aes(x=2800,y=2,hjust=1.05,vjust=0,label="400-800 BC - Most ancient evidence \n of african rice domesticated form"))+ geom_vline(aes(xintercept=estimate_mode(t.wild)))+ geom_text(aes(x=estimate_mode(t.wild)+10,y = 3,label="Mode of the distribution = 3201",hjust=-0.02,vjust=0)),
                    align = "v",
                    ncol = 1,
                    rel_heights = c(1,0.5,0.5),labels = c("A","B","C")),
                    filename = "/Users/cubry/Dropbox/Tracing the origin and diversity of african rice using 263 wild and cultivated depthly resequenced accessions/modelisation/graphes stairway et psmc_msmc/cult_vs_wild_psmc.pdf",
                    base_aspect_ratio = 0.8,
                    base_height = 15)

density_psmc <- plot_grid(psmc_wild + labs(x="") ,
                    density_cult + geom_ribbon(aes(x=c(2400,2800),ymin = 0,ymax=5),alpha=0.2)+
  geom_text(aes(x=2800,y=1,hjust=-0.02,vjust=0,label="400-800 BC - Most ancient evidence \n of african rice domesticated form"))+ geom_vline(aes(xintercept=estimate_mode(t.cult)))+ geom_text(aes(x=estimate_mode(t.cult)+10,y = 4.5,label="Mode of the distribution = 1847",hjust=1.05,vjust=0)),
                    density_wild +  geom_ribbon(aes(x=c(2400,2800),ymin = 0,ymax=3.5),alpha=0.2)+
  geom_text(aes(x=2800,y=2,hjust=1.05,vjust=0,label="400-800 BC - Most ancient evidence \n of african rice domesticated form"))+ geom_vline(aes(xintercept=estimate_mode(t.wild)))+ geom_text(aes(x=estimate_mode(t.wild)+10,y = 3,label="Mode of the distribution = 3201",hjust=-0.02,vjust=0)),
                    align = "v",
                    ncol = 1,
                    rel_heights = c(1,0.5,0.5),labels = c("B","C","D"))

pannel_msmc <- plot_grid(Ne_infered.combined_plot, density_psmc, labels = c('A', ''), ncol = 2, rel_heights = c(1.5, 1))

save_plot(pannel_msmc,
                    filename = "/Users/cubry/Dropbox/Tracing the origin and diversity of african rice using 263 wild and cultivated depthly resequenced accessions/modelisation/graphes stairway et psmc_msmc/pannel_psmc.pdf",
                    base_aspect_ratio = 2,
                    base_height = 10)
```

